### PR TITLE
DBaaS operator dependencies installation

### PR DESCRIPTION
**Pull request by redhatHameed**
## Description

The PR will install dependencies/components along with DBaaS:

MongoDB Atlas Operator
Crunchy Bridge Operator
dbaas-dynamic-plugin
console-telemetry-plugin
Screen Shot 2021-08-31 at 3 12 29 PM

Screen Shot 2021-08-31 at 3 12 52 PM

## Verification Steps

1. update the quay org/user name in the make file and catalog-source.yaml file with your org name
2. `make release`
3. `make catalog-update`
4. make sure the DBaaS operator installed 
5. make sure after successful installation of the DBaaS operator, it installed all its dependencies (CrunchBridge Operator, MongoDB atlas operator, DBaaS Dynamic Plugin<inside >, Telemetry Plugin)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist (Definition of Done)

- [x] All acceptance criteria specified in JIRA or in issue number have been completed
- [ ] Unit tests added that prove the fix is effective or the feature works 
- [ ] Documentation added for the feature
- [ ] all relevant tests are passing
- [ ] Code Review completed
- [ ] Verified independently by reviewer

### CODE REVIEW

File: api/v1alpha1/dbaasplatform_types.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+type PlatformsName string
+
+type PlatformsInstlnStatus string
+
+const (
+	CrunchyBridgeInstallation         PlatformsName = "crunchy-bridge"
+	MongoDBAtlasInstallation          PlatformsName = "mongodb-atlas"
+	DBassDynamicPluginInstallation    PlatformsName = "dbaas-dynamic-plugin"
+	Csv                               PlatformsName = "Csv"
+	ConsolTelemetryPluginInstallation PlatformsName = "console-telemetry-plugin"
```
Comment (jeremyary) [inline api/v1alpha1/dbaasplatform_types.go:32]: s/Consol/Console

File: controllers/reconcilers/constants.go [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+const (
+	INSTALL_NAMESPACE                        = "openshift-operators"
+	CATALOG_NAMESPACE                        = "openshift-marketplace"
+	MONGODB_ATLAS_CATLOG_IMG                 = "quay.io/ecosystem-appeng/mongodb-atlas-operator-catalog:0.6.25"
+	CRUNCHY_BRIDGE_CATLOG_IMG                = "quay.io/ahameed/crunchy-bridge-operator-catalog:v0.0.1"
+	DBAAS_DYNAMIC_PLUGIN_IMG                 = "quay.io/ecosystem-appeng/dbaas-dynamic-plugin:latest"
+	DBAAS_DYNAMIC_PLUGIN_NAME                = "dbaas-dynamic-plugin"
+	DBAAS_DYNAMIC_PLUGIN_NAMESPACE           = "dbaas-dynamic-plugin"
+	DBAAS_DYNAMIC_PLUGIN_DISPLAY_NAME        = "OpenShift DataBase as a Service Dynamic Plugin"
+	CONSOLE_TELEMETRY_PLUGIN_IMG             = "quay.io/christianvogt/console-telemetry-plugin:latest"
```
Comment (jeremyary) [inline controllers/reconcilers/constants.go:12]: Please pull this image and let's push it into our own team Quay org so we're protected against unexpected changes

File: controllers/reconcilers/constants.go [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+const (
+	INSTALL_NAMESPACE                        = "openshift-operators"
+	CATALOG_NAMESPACE                        = "openshift-marketplace"
+	MONGODB_ATLAS_CATLOG_IMG                 = "quay.io/ecosystem-appeng/mongodb-atlas-operator-catalog:0.6.25"
+	CRUNCHY_BRIDGE_CATLOG_IMG                = "quay.io/ahameed/crunchy-bridge-operator-catalog:v0.0.1"
+	DBAAS_DYNAMIC_PLUGIN_IMG                 = "quay.io/ecosystem-appeng/dbaas-dynamic-plugin:latest"
+	DBAAS_DYNAMIC_PLUGIN_NAME                = "dbaas-dynamic-plugin"
+	DBAAS_DYNAMIC_PLUGIN_NAMESPACE           = "dbaas-dynamic-plugin"
+	DBAAS_DYNAMIC_PLUGIN_DISPLAY_NAME        = "OpenShift DataBase as a Service Dynamic Plugin"
```
Comment (jeremyary) [inline controllers/reconcilers/constants.go:11]: s/DataBase/Database

File: controllers/reconcilers/crunchybridge_installation/crunchybridge_installation_reconciler.go [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+func NewReconciler(client client.Client, scheme *runtime.Scheme, logger logr.Logger) reconcilers.PlatformReconciler {
+	return &Reconciler{
+		client: client,
+		scheme: scheme,
+		logger: logger,
+	}
+}
+
+func (r *Reconciler) Reconcile(ctx context.Context, cr *v1.DBaaSPlatform, status2 *v1.DBaaSPlatformStatus) (v1.PlatformsInstlnStatus, error) {
+
```
Comment (jeremyary) [inline controllers/reconcilers/crunchybridge_installation/crunchybridge_installation_reconciler.go:34]: Are we checking to see if either Crunchy or Mongo are already installed on the cluster and skipping create/clean if our or another version of the provider operator are already present?

File: controllers/dbaasplatform_controller.go [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+		dbaasv1alpha1.CrunchyBridgeInstallation,
+		dbaasv1alpha1.MongoDBAtlasInstallation,
+		dbaasv1alpha1.DBassDynamicPluginInstallation,
+		dbaasv1alpha1.ConsolTelemetryPluginInstallation,
+	}
+
+}
+func (r *DBaaSPlatformReconciler) getCleanupPlatforms() []dbaasv1alpha1.PlatformsName {
+
+	return []dbaasv1alpha1.PlatformsName{
```
Comment (jeremyary) [inline controllers/dbaasplatform_controller.go:241]: we *may* need Service Binding Operator as well, trying to find out what I can on whether it's included on OSD or in a trusted source so we could use OLM dependency for it. Will update.

Comment (redhatHameed) [reply controllers/dbaasplatform_controller.go:243]: Added installation for service binding operators

File: controllers/reconcilers/constants.go (hunk 2 above)
Comment (redhatHameed) [reply controllers/reconcilers/constants.go]: Updated image org with our team org

**Comment by jianrongzhang89** [comment]
@redhatHameed When the dbaas-operator is deleted from OLM, should the crunchy bridge operator and mongodb atlas operator get deleted too? I tried in my test env, and found only the subscription and catalogsource objects were removed, but the operators are still there.

**Comment by jeremyary** [comment]
> should the crunchy bridge operator and mongodb atlas operator get deleted too?

This is acceptable for this PR & we'll discuss as a team soon.

**Review by jianrongzhang89 — COMMENTED** [review]
> > should the crunchy bridge operator and mongodb atlas operator get deleted too?
> 
> Provider operators not being removed is acceptable for this PR & we'll discuss as a team soon.

OK. It seems one can not even manually delete those provider operators with OLM when this happens.

**Comment by xieshenzh** [comment]
> if DBaaSPlatform CR is removed the DBaaS operator will clean all platforms (plugin+vendor operator), but in the other case when we remove the DBaaS operator it's not clean cleaning 100%. I have already set the owner reference as CSV but still, we need to find the best way to fix that.

It seems this happens when the `deployment` of the operator is terminated before the `platform` reconciler cleans up all the resources and remove the `finalizer` of the `platform` CR. If the operator is reinstalled, the cleanup will continue.
I am not sure if we can add the `finalizer` to the `deployment` of the operator instead of the `platform` CR, so that the `platform` reconciler can finish the cleanup.

File: config/manifests/bases/dbaas-operator.clusterserviceversion.yaml [hunk 6]
```diff
@@ -8,7 +8,7 @@ metadata:
     description: Import provider cloud instances for easy coupling with developer
       applications.
     operatorframework.io/suggested-namespace: openshift-dbaas-operator
-    operators.operatorframework.io/internal-objects: '["dbaasconnections.dbaas.redhat.com","dbaasproviders.dbaas.redhat.com","dbaastenants.dbaas.redhat.com","dbaasplatforms.dbaas.redhat.com"]'
+    operators.operatorframework.io/internal-objects: '["dbaasconnections.dbaas.redhat.com","dbaasproviders.dbaas.redhat.com","dbaastenants.dbaas.redhat.com"]'
```
Comment (xieshenzh) [inline config/manifests/bases/dbaas-operator.clusterserviceversion.yaml]: I think the `platform` CR is internal, and users don't need to see or create it from the UI.

File: config/manifests/bases/dbaas-operator.clusterserviceversion.yaml [hunk 7]
```diff
@@ -26,6 +26,11 @@ spec:
       kind: DBaaSInventory
       name: dbaasinventories.dbaas.redhat.com
       version: v1alpha1
+    - description: DBaaSPlatform is the Schema for the dbaasplatforms API
+      displayName: DBaa SPlatform
```
[the remark this answers is not in the document]
Comment (xieshenzh) [reply config/manifests/bases/dbaas-operator.clusterserviceversion.yaml:30]: Something like [this](https://github.com/RHEcosystemAppEng/dbaas-operator/blob/6639988367a762b0807827f73830e068b1215837/api/v1alpha1/dbaasconnection_types.go#L29) should work.

File: config/manifests/bases/dbaas-operator.clusterserviceversion.yaml (hunk 6 above)
Comment (jeremyary) [reply config/manifests/bases/dbaas-operator.clusterserviceversion.yaml]: +1

File: controllers/reconcilers/crunchybridge_installation/crunchybridge_installation_reconciler.go (hunk 4 above)
Comment (redhatHameed) [reply controllers/reconcilers/crunchybridge_installation/crunchybridge_installation_reconciler.go:34]: currently, we are not checking, and it's set auto-approval if there is a new version.

**Comment by redhatHameed** [comment]
> > if DBaaSPlatform CR is removed the DBaaS operator will clean all platforms (plugin+vendor operator), but in the other case when we remove the DBaaS operator it's not clean cleaning 100%. I have already set the owner reference as CSV but still, we need to find the best way to fix that.
> 
> It seems this happens when the `deployment` of the operator is terminated before the `platform` reconciler cleans up all the resources and remove the `finalizer` of the `platform` CR. If the operator is reinstalled, the cleanup will continue.
> I am not sure if we can add the `finalizer` to the `deployment` of the operator instead of the `platform` CR, so that the `platform` reconciler can finish the cleanup.

I think we can try this assuming that if we are not allowing to delete the CR to the user

File: config/manifests/bases/dbaas-operator.clusterserviceversion.yaml (hunk 7 above)
[the remark this answers is not in the document]
Comment (bmozaffa) [reply config/manifests/bases/dbaas-operator.clusterserviceversion.yaml:30]: Nice, this bugged me and I couldn't find anything that worked. Please try this, @jianrongzhang89 !

[the remark this answers is not in the document]
Comment (redhatHameed) [reply config/manifests/bases/dbaas-operator.clusterserviceversion.yaml:30]: this worked for me. @bmozaffa

[the remark this answers is not in the document]
Comment (jeremyary) [reply config/manifests/bases/dbaas-operator.clusterserviceversion.yaml:30]: Fantastic... now, can we use this trick to relabel our DBaaSInventory as "Provider Account" to align with UXD as well? This is also on our list to do... :)

[the remark this answers is not in the document]
Comment (redhatHameed) [reply config/manifests/bases/dbaas-operator.clusterserviceversion.yaml]: updated in the this PR.

File: config/manifests/bases/dbaas-operator.clusterserviceversion.yaml (hunk 6 above)
Comment (redhatHameed) [reply config/manifests/bases/dbaas-operator.clusterserviceversion.yaml]: done

File: controllers/dbaasplatform_controller.go [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+		}
+		// error fetching DBaaSPlatform instance, requeue and try again
+		log.Error(err, "Error in Get of DBaaSPlatform CR")
+		return ctrl.Result{}, err
+	}
+
+	if cr.DeletionTimestamp == nil && len(cr.Finalizers) == 0 {
+		cr.Finalizers = append(cr.Finalizers, DBaaSPlatformFinalizer)
+		err = r.Update(ctx, cr)
+		return ctrl.Result{}, err
```
[the remark this answers is not in the document]
Comment (redhatHameed) [reply controllers/dbaasplatform_controller.go:96]: Here it will reconcile again as we are updating the CR

**Comment by xieshenzh** [comment]
> > > if DBaaSPlatform CR is removed the DBaaS operator will clean all platforms (plugin+vendor operator), but in the other case when we remove the DBaaS operator it's not clean cleaning 100%. I have already set the owner reference as CSV but still, we need to find the best way to fix that.
> > 
> > 
> > It seems this happens when the `deployment` of the operator is terminated before the `platform` reconciler cleans up all the resources and remove the `finalizer` of the `platform` CR. If the operator is reinstalled, the cleanup will continue.
> > I am not sure if we can add the `finalizer` to the `deployment` of the operator instead of the `platform` CR, so that the `platform` reconciler can finish the cleanup.
> 
> I think we can try this assuming that if we are not allowing to delete the CR to the user

It doesn't work. `Pod` and `Deployment` are terminated even with `finalizer`. `Finalizer` only delays the deletion of the resources.

File: controllers/reconcilers/crunchybridge_installation/crunchybridge_installation_reconciler.go [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+		if deployment.Name == "crunchy-bridge-operator-controller-manager" {
+			if deployment.Status.ReadyReplicas > 0 {
+				return v1.ResultSuccess, nil
+			}
+		}
+	}
+	return v1.ResultInProgress, nil
+}
+
+func GetCrunchyBridgeSubscription() *v1alpha1.Subscription {
```
[the remark this answers is not in the document]
Comment (redhatHameed) [reply controllers/reconcilers/crunchybridge_installation/crunchybridge_installation_reconciler.go]: updated

**Comment by xieshenzh** [comment]
Once the dbaas operator is uninstalled, the `pod` will be terminated. Sometimes, the `dbaasplatform` reconciler won't be able to finish the uninstallation of the dependencies before the termination of the `pod`. 
As a result, the `dbaasplatform` CR will still have deletion timestamp and finalizer, which means it will not be deleted. And if the operator is reinstalled, the `dbaasplatform` reconciler will continue the uninstallation.

I am not sure how to solve this problem, if we want to use the dbaas operator to uninstall the dependencies. The dbaas operator could be terminated before it finishes the uninstallation. 

@jeremyary did we miss something? Does the observability operator do something differently to handle this scenario?

**Comment by bmozaffa** [comment]
@xieshenzh , I may be missing something obvious here, but can't we have a finalizer on the dbaas operator that is only removed after it successfully finishes the cleanup?

**Comment by xieshenzh** [comment]
> @xieshenzh , I may be missing something obvious here, but can't we have a finalizer on the dbaas operator that is only removed after it successfully finishes the cleanup?

I have tried to add a finalizer on the `deployment` and `pod` of the dbaas operator.
After uninstalling the dbaas operator, if the `deployment` has a finalizer, the `pod` will be deleted anyway and the `deployment` will have 0 pods.
If the `pod` has a finalizer, it will still stop running, but it will not be deleted until the finalizer is removed. It will be in a "terminating" state.
So in both cases, the operator will stop running.

Maybe I should try to add a finalizer to the CSV of the opeartor. The CSV is the owner of the `dbaasplatform` CR that we need to delete in order to start the cleanup. I'll try and see if this will work.

File: controllers/dbaasplatform_controller.go [hunk 10]
```diff
@@ -0,0 +1,297 @@
...
+	if err != nil {
+		return nil, err
+	}
+	dbaaSPlatformList := &dbaasv1alpha1.DBaaSPlatformList{}
+	listOpts := []k8sclient.ListOption{
+		k8sclient.InNamespace(namespace),
+	}
+	err = serverClient.List(ctx, dbaaSPlatformList, listOpts...)
+	if err != nil {
+		return nil, fmt.Errorf("Could not get a list of dbaas platform intallation CR: %w", err)
```
Comment (jianrongzhang89) [inline controllers/dbaasplatform_controller.go:191]: "Could not" => "could not", by convention.

**Review by jianrongzhang89 — COMMENTED** [review]
> > @xieshenzh , I may be missing something obvious here, but can't we have a finalizer on the dbaas operator that is only removed after it successfully finishes the cleanup?
> 
> I have tried to add a finalizer on the `deployment` and `pod` of the dbaas operator.
> After uninstalling the dbaas operator, if the `deployment` has a finalizer, the `pod` will be deleted anyway and the `deployment` will have 0 pods.
> If the `pod` has a finalizer, it will still stop running, but it will not be deleted until the finalizer is removed. It will be in a "terminating" state.
> So in both cases, the operator will stop running.
> 
> Maybe I should try to add a finalizer to the CSV of the opeartor. The CSV is the owner of the `dbaasplatform` CR that we need to delete in order to start the cleanup. I'll try and see if this will work.

@xieshenzh @redhatHameed @jeremyary what about setting the owner references of all dependent objects (catalog source and subscription objects etc) to the DBaaS Operator's CSV? In this way the cleanup becomes automatic. I will try and see how it goes.

**Comment by xieshenzh** [comment]
I have tried to add `finalizer` on `csv` and `subscription`. In both cases, the `pod` will be terminated once I uninstall the operator.
@jeremyary is this because I am uninstalling the operator in OLM? Will the behavior be different on OSD?

**Comment by xieshenzh** [comment]
> @xieshenzh @redhatHameed @jeremyary what about setting the owner references of all dependent objects (catalog source and subscription objects etc) to the DBaaS Operator's CSV? In this way the cleanup becomes automatic. I will try and see how it goes.

We have to delete some cluster scoped resources, and update some resources instead of deleting them.

**Comment by redhatHameed** [comment]
> I have tried to add `finalizer` on `csv` and `subscription`. In both cases, the `pod` will be terminated once I uninstall the operator.
> @jeremyary is this because I am uninstalling the operator in OLM? Will the behavior be different on OSD?

on OSD, the operator gets installed using the addon: and I am not sure if we can uninstall/delete the addon because it's automatically managed. 
that being said we can uninstall the operator on OSD? @jeremyary

**Comment by redhatHameed** [comment]
> @xieshenzh , I may be missing something obvious here, but can't we have a finalizer on the dbaas operator that is only removed after it successfully finishes the cleanup?

we are facing the kind of same issue
https://github.com/operator-framework/operator-lifecycle-manager/issues/1302

File: controllers/reconcilers/crunchybridge_installation/crunchybridge_installation_reconciler.go (hunk 4 above)
Comment (redhatHameed) [reply controllers/reconcilers/crunchybridge_installation/crunchybridge_installation_reconciler.go:34]: @jeremyary let me know if needs any action for that thanks

**Comment by redhatHameed** [comment]
@jianrongzhang89 @jeremyary can you have another look or approve this PR thanks

**Review by jeremyary — APPROVED** [review]

**Pull request merged**